pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -150,11 +165,47 @@ func (pl *DefaultPreemption) preempt(ctx context.Context, state *framework.Cycle
 	return bestCandidate.Name(), nil
 }
 
+// calculateNumCandidates returns the number of candidates the FindCandidates
+// method must produce from dry running based on the constraints given by
+// <minCandidateNodesPercentage> and <minCandidateNodesAbsolute>. The number of
+// candidates returned will never be greater than <numNodes>.
+func (pl *DefaultPreemption) calculateNumCandidates(numNodes int32) int32 {
+	n := (numNodes * pl.args.MinCandidateNodesPercentage) / 100
+	if n < pl.args.MinCandidateNodesAbsolute {
+		n = pl.args.MinCandidateNodesAbsolute
+	}
+	if n > numNodes {
+		n = numNodes
+	}
+	return n
+}
+
+// updateAndReturnOffset updates its internal state returns an <offset> and
+// <numCandidates> that the caller must use to find candidates. A moving offset
+// is used as the starting index when dry running preemption so that the same
+// set of nodes don't get preempted every time. If <offset> + <numCandidates>
+// runs past the number of nodes, the caller is expected to wrap around the
+// node list.
+func (pl *DefaultPreemption) updateAndReturnOffset(numNodes int32) (int32, int32) {
+	offset := pl.candidateOffset
+	numCandidates := pl.calculateNumCandidates(numNodes)
+	if pl.candidateOffset >= numNodes {
+		// This happens when the cluster is downscaled and candidateOffset is past
+		// the new, reduced number of nodes. Reset the counter and restart from the
+		// beginning.
+		pl.candidateOffset = 0
+		offset = 0
+	} else {
+		pl.candidateOffset = (pl.candidateOffset + numCandidates) % numNodes
+	}
+	return offset, numCandidates
+}
+
 // FindCandidates calculates a slice of preemption candidates.
 // Each candidate is executable to make the given <pod> schedulable.
-func FindCandidates(ctx context.Context, cs kubernetes.Interface, state *framework.CycleState, pod *v1.Pod,
-	m framework.NodeToStatusMap, ph framework.PreemptHandle, nodeLister framework.NodeInfoLister,
-	pdbLister policylisters.PodDisruptionBudgetLister) ([]Candidate, error) {
+func FindCandidates(ctx context.Context, cs kubernetes.Interface,
```

Comment:
`func (pl *DefaultPreemption)`  instead of passing it as an argument.

Reply:
ph, nodeLister and pdbLister can all be accessed from pl as well.

Reply:
sorry, `cs` as well.

Reply:
Ah, I missed that! Done.